db/index_changes.go

```diff
@@ -403,17 +422,105 @@ func (db *Database) initializeChannelFeeds(channelsSince channels.TimedSet, opti
 	return feeds, nil
 }
 
+// Calculates the range for backfill processing, for compatibility with changes streaming.
+//  For the given:
+//    - since clock SinceClock
+//    - current backfill position vb-B.seq-B,
+//    - granting sequence vb-G.seq-G
+//  We want to return everything for the channel from vb-B.seq-B up to vb-G.seq-G as backfill,
+//  and also return everything from vb-G.seq-G to vb-max.seq-max that's earlier the since value,
+//
+// From Clock:
+// If vb < vb-B, seq = MaxUint
+// If vb = vb-B, seq = seq-b
+// if vb > vb-B, seq = SinceClock seq
+// To Clock:
+// If vb < vb-G, seq = MaxInt
+// If vb = vb-G, seq = seq-G
+// if vb > vb-G, seq = SinceClock
+
+func calculateBackfillRange(backfillPosition base.VbSequence, triggerPosition base.VbSequence, sinceClock base.SequenceClock) (fromClock, toClock base.SequenceClock) {
+
+	fromClock = base.NewSequenceClockImpl()
+	toClock = base.NewSequenceClockImpl()
+	MAX_SEQUENCE := uint64(math.MaxUint64)
+
+	for vbInt, _ := range sinceClock.Value() {
+		vbNo := uint16(vbInt)
+		var fromSeq, toSeq uint64
+
+		// Calculate from sequence for vbucket
+		if vbNo < backfillPosition.Vb {
+			fromSeq = MAX_SEQUENCE
+		} else if vbNo > backfillPosition.Vb {
+			fromSeq = sinceClock.GetSequence(vbNo)
+		} else { // vbNo == backfillPosition.Vb
+			fromSeq = backfillPosition.Seq
+		}
+		fromClock.SetSequence(vbNo, fromSeq)
+
+		// Calculate to sequence for vbucket
+		if vbNo < triggerPosition.Vb {
+			toSeq = MAX_SEQUENCE
+		} else if vbNo > triggerPosition.Vb {
+			toSeq = sinceClock.GetSequence(vbNo)
+		} else { // vbNo==triggerPosition.Vb
+			toSeq = triggerPosition.Seq
+		}
+		toClock.SetSequence(vbNo, toSeq)
+	}
+
+	return fromClock, toClock
+}
+
 // Creates a Go-channel of all the changes made on a channel.
 // Does NOT handle the Wait option. Does NOT check authorization.
 func (db *Database) vectorChangesFeed(channel string, options ChangesOptions) (<-chan *ChangeEntry, error) {
 	dbExpvars.Add("channelChangesFeeds", 1)
-	log, err := db.changeCache.GetChanges(channel, options)
-	base.LogTo("Changes+", "[changesFeed] Found %d changes for channel %s", len(log), channel)
-	if err != nil {
-		return nil, err
+	changeIndex, ok := db.changeCache.(*kvChangeIndex)
+	if !ok {
+		return nil, errors.New("Called vectorChangesFeed with non-index cache.")
 	}
 
-	if len(log) == 0 {
+	// If we're in backfill for this channel, we make one reader call for the backfill and one for non-backfill.  Without
+	// two requests, it's not possible to use limit to prevent a full index scan of some vbuckets while still preserving ordering.
+	// The second call should only be made if the first request doesn't return limit changes.
+	var backfillLog, log []*LogEntry
```

Comment:
this is a bit hard to mentally parse despite being go code .. 

I know it's one extra line but think this would be a bit more readable:

```
	var (
		backfillLog []*LogEntry  // All the changes in this backfill
		log []*LogEntry // The entire changeset in the database (???)
	 	err error
	)	

```